src/Data/Nat/DivMod.agda

```diff
@@ -6,143 +6,115 @@
+-- Certified operations (i.e. operations with proofs)
 
-  division-lemma mod-acc div-acc (suc d) (suc n) = begin
+-- A specification of integer division.
 
-    mod-acc + div-acc * suc s + suc d                   ≡⟨ solve 4
-                                                                 (λ mod-acc div-acc n d →
-                                                                      mod-acc :+ div-acc :* (con 1 :+ (mod-acc :+ (con 1 :+ n))) :+ (con 1 :+ d)
-                                                                        :=
-                                                                      con 1 :+ mod-acc :+ div-acc :* (con 2 :+ mod-acc :+ n) :+ d)
-                                                                 P.refl mod-acc div-acc n d ⟩
-    suc mod-acc + div-acc * suc s′ + d                  ≡⟨ division-lemma (suc mod-acc) div-acc d n ⟩
+record DivMod (dividend divisor : ℕ) : Set where
+  constructor result
+  field
+    quotient  : ℕ
+    remainder : Fin divisor
+    property  : dividend ≡ toℕ remainder + quotient * divisor
 
-    mod-helper (suc mod-acc) s′ d n +
-    div-helper      div-acc  s′ d n * suc s′            ≡⟨ P.cong (λ s → mod-helper (suc mod-acc) s d n +
-                                                                         div-helper div-acc s d n * suc s)
-                                                                  (P.sym (+-suc mod-acc n)) ⟩
+infixl 7 _mod_ _divMod_
 
-    mod-helper (suc mod-acc) s d n +
-    div-helper      div-acc  s d n * suc s              ≡⟨⟩
+-- Certified modulus
 
-    mod-helper      mod-acc  s (suc d) (suc n) +
-    div-helper      div-acc  s (suc d) (suc n) * suc s  ∎
+_mod_ : (dividend divisor : ℕ) {≢0 : False (divisor ≟ 0)} → Fin divisor
+(a mod 0) {}
+(a mod suc n) = fromℕ≤″ (a % suc n) (Nat.erase (≤⇒≤″ (a%n<n a n)))
 
-    where
-    s  = mod-acc + suc n
-    s′ = suc mod-acc + n
+-- Returns modulus and division result with correctness proof
 
 _divMod_ : (dividend divisor : ℕ) {≢0 : False (divisor ≟ 0)} →
            DivMod dividend divisor
-(d divMod 0) {}
-(d divMod suc s) =
-  result (d div suc s) (d mod suc s) (TrustMe.erase (begin
-    d                                                        ≡⟨ division-lemma 0 0 d s ⟩
-    mod-helper 0 s d s         + div-helper 0 s d s * suc s  ≡⟨ P.cong₂ _+_ (P.sym (FinP.toℕ-fromℕ≤ lemma)) P.refl ⟩
-    toℕ (Fin.fromℕ≤    lemma)  + div-helper 0 s d s * suc s  ≡⟨ P.cong (λ n → toℕ n + div-helper 0 s d s * suc s)
-                                                                       (FinP.fromℕ≤≡fromℕ≤″ lemma _) ⟩
-    toℕ (Fin.fromℕ≤″ _ lemma′) + div-helper 0 s d s * suc s  ∎))
+(a divMod 0) {}
+(a divMod suc n) = result (a div suc n) (a mod suc n)
+  (TrustMe.erase (begin
+    a                                                ≡⟨ a≡a%n+[a/n]*n a n ⟩
+    (a % suc n)             + (a div suc n) * suc n  ≡⟨ cong₂ _+_ (sym (toℕ-fromℕ≤″ lemma′)) refl ⟩
+    toℕ (fromℕ≤″ _ lemma′) + (a div suc n) * suc n  ∎))
   where
```

Comment:
Is it worth naming `a div suc n` here? It's used 3 times (once in result + twice in the proof).
I would also write `$ TrustMe.erase $ begin` instead of the parentheses around the big block.